[WIP] Added canceled field to unified_jobs model

thedoubl3j opened the pull request:
- When a job is canceled, finished will now return a null and the canceled field will populate with date/time. related #3988 

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->

 - Bugfix Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API


##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
make VERSION
awx: 9.0.1

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
Before, this would look like 
<!--- Paste verbatim command output below, e.g. before and after your change -->
```
    "scm_type": "git",
    "scm_url": "https://github.com/ansible/test-playbooks.git",
    "scm_branch": "",
    "scm_clean": false,
    "scm_delete_on_update": false,
    "credential": null,
    "timeout": 0,
    "unified_job_template": 3838,
    "launch_type": "manual",
    "status": "canceled",
    "failed": true,
    "started": null,
    "finished": "2019-06-03T12:39:05.272329Z",
    "elapsed": 0.0,
    "job_args": "",
    "job_cwd": "",
    "job_env": {},
    "job_explanation": "",
    "execution_node": "",
    "result_traceback": "",
    "event_processing_finished": true,
    "project": 3838,
    "job_type": "check",
```
but now, the api returns a null for `finished:` and also have a canceled field. 

```
    "created": "2019-11-26T16:13:20.282945Z",
    "modified": "2019-11-26T16:13:25.286318Z",
    "name": "test_the_thing",
    "description": "",
    "job_type": "run",
    "inventory": 1,
    "project": 8,
    "playbook": "check_postgres.yml",
    "scm_branch": "",
    "forks": 0,
    "limit": "",
    "verbosity": 0,
    "extra_vars": "{}",
    "job_tags": "",
    "force_handlers": false,
    "skip_tags": "",
    "start_at_task": "",
    "timeout": 0,
    "use_fact_cache": false,
    "unified_job_template": 9,
    "launch_type": "relaunch",
    "status": "canceled",
    "failed": true,
    "started": "2019-11-26T16:13:25.379121Z",
    "finished": null,
    "canceled": "2019-11-26T16:13:31.504482Z",
    "elapsed": 21.128626,
```

thedoubl3j committed:
Added canceled_on field to unified_jobs model
  - When a job is canceled, new field canceled_on will populate with the timestamp of when it was canceled based on cancel_flag

thedoubl3j committed:
Added canceled field to unified_jobs model
  - When a job is canceled, finished will now return a null and the canceled field will populate with date/time

thedoubl3j commented:
big props to @rooftopcellist for all of the help getting stuff re-organized and explaining parts of the system that I would have no idea how to look at on my own.

ryanpetrello reviewed (changes requested):
This seems like duplicate information - does filtering on `status=canceled` not work?

ryanpetrello commented:
Oh, I see, you want the *time* it was canceled?

ryanpetrello commented:
<hrm>

I want to think on this one :)

ryanpetrello commented on awx/main/migrations/0100_v370_unifiedjob_canceled.py:
```diff
@@ -0,0 +1,18 @@
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('main', '0099_v370_projectupdate_job_tags'),
```
This file probably needs to be `0101_v370`, we've added a `0100` at this point.

jbradberry replied:
@thedoubl3j drop me a line and I can link you to my git rebase script that will renumber migrations for you.

thedoubl3j replied:
+1 @jbradberry definitely. plan on picking back up the work on this either tomorrow or thursday

AlanCoding commented on awx/main/models/unified_jobs.py:
```diff
@@ -810,11 +817,18 @@ def save(self, *args, **kwargs):
+            # Record the `finished` time and if the job was canceled, set it to null.
             self.finished = now()
+            if self.status == 'canceled':
+                self.finished = None
```
If it reports being both finished and canceled, I'd rather know than to lose that information here.

AlanCoding replied:
this is starting to make sense to me, in that there are some cases where we should desire to not set `finished`.

However, I would try to do that in a more targeted way. Specifically, if `started` datetime field is None, and status is canceled, then `finished` should not be set.

AlanCoding commented on awx/main/models/unified_jobs.py:
```diff
@@ -810,11 +817,18 @@ def save(self, *args, **kwargs):
+                self.canceled = now()
             if 'finished' not in update_fields:
                 update_fields.append('finished')
+            if 'canceled' not in update_fields:
```
You need to gate this conditional with and additional condition like `and not self.finished`, like we do for the finished field. It probably makes sense to move some of that logic for the finished event too.

thedoubl3j replied:
What would be an example that you would want to see @AlanCoding? Is the gate on line 822 not sufficient?

AlanCoding replied:
If you only want this to run when the `finished` time is set, then I don't really follow what your objective is. That would make it a purely aesthetic change. The client can deduce all of this information now, with the knowledge that `finished` is roughly the time that the subprocess stops, whether it was canceled or not.

AlanCoding commented on awx/main/models/unified_jobs.py:
```diff
@@ -810,11 +817,18 @@ def save(self, *args, **kwargs):
         # completion time, and record its output to the database.
+        #import sdb
+        #sdb.set_trace()
         if self.status in ('successful', 'failed', 'error', 'canceled') and not self.finished:
```
The time that `status` changes to canceled is not the most important thing to me. I believe that will be at least partially duplicated information (particularly with finished).

What I think people should want is not the time that the job _actually_ cancels, but the time of the cancel request. For that, you want to capture the moment when the `cancel_flag` flips. That is very useful information because it tracks a user action, and can be used to diagnose situations where the system behavior... doesn't quite match what the user requested.

thedoubl3j commented:
@ryanpetrello and or anyone else @rooftopcellist. Spoke with @AlanCoding yesterday and we decided to this PR a different way and remove the canceled field altogether, especially since it was essentially the finished time when the smoke cleared. 

We were thinking of adding a field called `canceld_on` that would just be the time the post for cancel was sent and still report the finished time (because it technically still can) and just adding another logic block that checks to see if the cancel_flag is true. 

I will be picking this back up I hope either tomorrow or thursday and will report back with the current workings.

keithjgrant committed:
Start inventory detail

* Create VariablesDetail for read-only variables view
* Sketch out InventoryDetail
* Create CardBody and TabbedCardHeader for common custom styling

jainnikhil30 committed:
fix the saml_admin_attr not working

ryanpetrello committed:
only restrict OAuth2 tokens for external accounts *at creation time*

related: https://github.com/ansible/awx/pull/5477

keithjgrant committed:
finish InventoryDetail

keithjgrant committed:
add tests for VariablesDetail & InventoryDetail

keithjgrant committed:
switch all tabbed screens to use TabbedCardHeader

keithjgrant committed:
use VariablesDetail for displaying variables field in details views

keithjgrant committed:
use UserDateDetail in OrganizationDetail

thedoubl3j committed:
deleted migrations file and testing things

keithjgrant committed:
InventoryDetail: handle content loading state & errors better

keithjgrant committed:
Fix UserDateDetail translation

Add UserDateDetail to Org detail & InventoryGroupDetail
Add VariablesDetail to InventoryGroupDetail

marshmalien committed:
Add Credential List and unit tests

keithjgrant committed:
switch to our CardBody component on all screens, for consistent padding/spacing

marshmalien committed:
Fetch credential types and options only on initial render

marshmalien committed:
Use credential_types from credential.summary_fields to display "Type" column

keithjgrant committed:
Add Inventory detail edit button

keithjgrant committed:
add delete button to InventoryDetail

marshmalien committed:
Reset selected list after successful delete

keithjgrant committed:
update tests for detail view changes

thedoubl3j committed:
Merge branch 'devel' of github.com:ansible/awx into canceled_jobs

keithjgrant committed:
de-lint

keithjgrant committed:
fix inventory saving without insights credential specified

AlanCoding commented on awx/main/models/unified_jobs.py:
```diff
@@ -1359,6 +1373,11 @@ def cancel(self, job_explanation=None, is_chain=False):
+                    # Record the `finished` or 'canceled' time.
+                    self.canceled_on = now()
+                    if 'canceled_on' not in update_fields:
+                        update_fields.append('canceled_on')
```
this looks good to me

AlanCoding commented on awx/main/models/unified_jobs.py:
```diff
@@ -810,11 +817,18 @@ def save(self, *args, **kwargs):
         # Sanity check: Has the job just completed? If so, mark down its
         # completion time, and record its output to the database.
+        #import sdb
+        #sdb.set_trace()
```
remove

AlanCoding commented on awx/main/models/unified_jobs.py:
```diff
@@ -810,11 +817,18 @@ def save(self, *args, **kwargs):
             self.finished = now()
+            if self.status == 'canceled':
+                self.finished = None
+                self.canceled = now()
```
I don't think `canceled` is a field anymore?

thedoubl3j replied:
oh dang it man, I pushed. This is still a WIP

thedoubl3j replied:
sorry alan, should have closed before I pushed, this was a rebase. still need to clean a lot of stuff up.

softwarefactory-project-zuul[bot] committed:
Merge pull request #5520 from keithjgrant/5261-inventory-detail-b

Inventory Detail

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

softwarefactory-project-zuul[bot] committed:
Merge pull request #5542 from marshmalien/cred-list

Add Credential List 

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

keithjgrant committed:
fix form validation for Organization select in Inventory form

wenottingham committed:
Fix survey validation to always retun an error code if erroring

softwarefactory-project-zuul[bot] committed:
Merge pull request #5560 from wenottingham/bad-request-bad-naughty-evil-request

Fix survey validation to always retun an error code if erroring

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

beeankha committed:
Enable easier building of playbooks for local Collections module testing

softwarefactory-project-zuul[bot] committed:
Merge pull request #5563 from beeankha/update_makefile

Edit Makefile for Easier Collections Building/Playbook Testing

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

loitho committed:
Add a uwsgi param to prevent SAML error

Add the uwsgi_param 'HTTP_X_FORWARDED_PORT' to nginx configuration,
This prevents the python-saml "invalid_response" error

related issue : #5570 and #1016

Signed-off-by: loitho

AlanCoding committed:
Revert "Apply migration flag check to task manager"

This reverts commit a0910eb6de3c4713199e73c43f003c48c2895f9f.

AlanCoding committed:
Revert "Reduce API response times by caching migration flag"

This reverts commit 5433af6716dd82cf206a7ebec8b9d6e0a4592889.

softwarefactory-project-zuul[bot] committed:
Merge pull request #5499 from ryanpetrello/more-oauth-tinkering

only restrict OAuth2 tokens for external accounts *at creation time*

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

softwarefactory-project-zuul[bot] committed:
Merge pull request #5579 from AlanCoding/revert_migration_flag

Revert migration flag

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

mabashian committed:
Upgrades pf deps to latest.  Adds ability to click on row items in order to select them rather than having to click on checkboxes/radio buttons.

AlanCoding committed:
Remove incorrect activity stream entries related to managed types

softwarefactory-project-zuul[bot] committed:
Merge pull request #5581 from AlanCoding/hush_managed_types

Remove incorrect activity stream entries related to managed types

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

softwarefactory-project-zuul[bot] committed:
Merge pull request #5497 from jainnikhil30/devel

fix the saml_admin_attr not working

Reviewed-by: Ryan Petrello
             https://github.com/ryanpetrello

softwarefactory-project-zuul[bot] committed:
Merge pull request #5577 from loitho/devel

Add a uwsgi param to prevent SAML error

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

ryanpetrello committed:
add the ability to generate dot graphs for per-request profiling

softwarefactory-project-zuul[bot] committed:
Merge pull request #5593 from ryanpetrello/dot-dot-dot

add the ability to generate dot graphs for per-request profiling

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

softwarefactory-project-zuul[bot] committed:
Merge pull request #5553 from keithjgrant/inventory-add-save

Inventory Add form fixes

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

softwarefactory-project-zuul[bot] committed:
Merge pull request #5580 from mabashian/upgrade-pf-deps

Upgrades pf deps to latest

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

AlexSCorey committed:
Adds Promote Delete Modal To InventoryGroupDetails

It also removes a comment/reminder to remove some code after CredentialsLookUp
refactor was completed.  Now that it has been completed that code has been removed.

softwarefactory-project-zuul[bot] committed:
Merge pull request #5605 from AlexSCorey/5494-IncorrectInvGroupDetailDeleteModal

Adds Promote Delete Modal To InventoryGroupDetails

Reviewed-by: https://github.com/apps/softwarefactory-project-zuul

exicx committed:
change set-context to use-context

set-context allows setting configuration within a provided context, to change contexts we need "use-context"

$ kubectl config
...
  set-context     Sets a context entry in kubeconfig
  use-context     Sets the current-context in a kubeconfig file

softwarefactory-project-zuul[bot] committed:
Merge pull request #5607 from exicx/exicx-patch-context

change set-context to use-context

Reviewed-by: Shane McDonald <me@shanemcd.com>
             https://github.com/shanemcd

thedoubl3j committed:
add canceled_on field and migration file

thedoubl3j committed:
Merge branch 'canceled_jobs' of github.com:thedoubl3j/awx into canceled_jobs

thedoubl3j commented:
closing while i clean up this git mess so that CI doesn't run
